src/devices/gropdf/pdfmom.1.man
.TH pdfmom @MAN1EXT@ "@MDATE@" "groff @VERSION@"
.SH Name
pdfmom \- produce PDF documents using the
.I mom
macro package for
.I groff
.
.
.\" ====================================================================
.\" Legal Terms
.\" ====================================================================
.\"
.\" Copyright (C) 2012-2020 Free Software Foundation, Inc.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of
.\" this manual under the conditions for verbatim copying, provided that
.\" the entire resulting derived work is distributed under the terms of
.\" a permission notice identical to this one.
.\"
.\" Permission is granted to copy and distribute translations of this
.\" manual into another language, under the above conditions for
.\" modified versions, except that this permission notice may be
.\" included in translations approved by the Free Software Foundation
.\" instead of in the original English.
.
.
.\" Save and disable compatibility mode (for, e.g., Solaris 10/11).
.do nr *groff_pdfmom_1_man_C \n[.cp]
.cp 0
.
.\" Define fallback for groff 1.23's MR macro if the system lacks it.
.nr do-fallback 0
.if !\n(.f           .nr do-fallback 1 \" mandoc
.if  \n(.g .if !d MR .nr do-fallback 1 \" older groff
.if !\n(.g           .nr do-fallback 1 \" non-groff *roff
.if \n[do-fallback]  \{\
.  de MR
.    ie \\n(.$=1 \
.      I \%\\$1
.    el \
.      IR \%\\$1 (\\$2)\\$3
.  .
.\}
.rr do-fallback
.
.
.\" ====================================================================
.SH Synopsis
.\" ====================================================================
.
.SY pdfmom
.RB [ \-Tpdf ]
.RI [ groff-options ]
.RI [ file\~ .\|.\|.]
.YS
.
.
.SY pdfmom
.B \-Tps
.RI [ pdfroff-options ]
.RI [ groff-options ]
.RI [ file\~ .\|.\|.]
.YS
.
.
.SY pdfmom
.B \-v
.
.SY pdfmom
.B \%\-\-version
.YS
.
.
.\" ====================================================================
.SH Description
.\" ====================================================================
.
.I pdfmom
is a wrapper around
.MR groff @MAN1EXT@
that facilitates the production of PDF documents from files
formatted with the
.I mom
macros.
.
.
.P
.I pdfmom
prints to the standard output,
so output must usually be redirected to a destination file.
.
The size of the final PDF can be reduced by piping the output
through
.MR ps2pdf 1 .
.
.
.P
If called with the
.B \-Tpdf
option (which is the default),
.I pdfmom
processes files using
.IR groff 's
native PDF driver,
.MR gropdf @MAN1EXT@ .
.
If
.B \-Tps
is given,
processing is passed over to
.IR pdfroff ,
which uses
.IR groff 's
PostScript driver.
.
In either case,
multiple runs of the source file are performed in order to satisfy any
forward references in the document.
.
.
.P
.I pdfmom
accepts all the same options as
.IR groff .
.
If
.B \-Tps
is given,
the options associated with
.I pdfroff
are accepted as well.
.
When
.I pdfmom
calls
.IR pdfroff ,
the options
.RB \[lq] "\-mpdfmark \-mom \-\-no\-toc" \[rq]
options are implied and should not be given on the command line.
.
Equally,
it is not necessary to supply the
.B \-mom
or
.B "\-m\~mom"
options when
.B \-Tps
is absent.
.
.
.P
PDF integration with the
.I mom
macros is discussed in full in the manual
\[lq]Producing PDFs with
.I groff
and
.IR mom \[rq],
which was itself produced with
.IR pdfmom .
.
.
.P
If called with the
.B \-v
or
.B \%\-\-version
options,
.I pdfmom
displays its version information and exits.
.
.
.\" ====================================================================
.SH Authors
.\" ====================================================================
.
.I pdfmom
was written by
.MT deri@\:chuzzlewit\:.myzen\:.co\:.uk
Deri James
.ME
and
.MT peter@\:schaffter\:.ca
Peter Schaffter
.ME ,
and is maintained by James.
.
.
.\" ====================================================================
.SH "See also"
.\" ====================================================================
.
.TP
.I @PDFDOCDIR@/\:mom\-pdf.pdf
\[lq]Producing PDFs with
.I groff
and
.IR mom \[rq],
by Deri James and Peter Schaffter.
.
This file,
together with its source,
.IR mom\-pdf.mom ,
is part of the
.I groff
distribution.
.
.
.P
.MR groff @MAN1EXT@ ,
.MR gropdf @MAN1EXT@ ,
.MR pdfroff @MAN1EXT@ ,
.MR ps2pdf 1
.
.
.\" Restore compatibility mode (for, e.g., Solaris 10/11).
.cp \n[*groff_pdfmom_1_man_C]
.do rr *groff_pdfmom_1_man_C
.
.
.\" Local Variables:
.\" fill-column: 72
.\" mode: nroff
.\" End:
.\" vim: set filetype=groff textwidth=72:


src/devices/gropdf/pdfmom.pl
#!@PERL@
#
#	pdfmom		: Frontend to run groff -mom to produce PDFs
#	Deri James	: Friday 16 Mar 2012
#

# Copyright (C) 2012-2020 Free Software Foundation, Inc.
#      Written by Deri James <deri@chuzzlewit.myzen.co.uk>
#
# This file is part of groff.
#
# groff is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free
# Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# groff is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;
use File::Temp qw/tempfile/;
my @cmd;
my $dev='pdf';
my $preconv='';
my $readstdin=1;
my $RT_SEP='@RT_SEP@';

$ENV{PATH}=$ENV{GROFF_BIN_PATH}.$RT_SEP.$ENV{PATH} if exists($ENV{GROFF_BIN_PATH});
$ENV{TMPDIR}=$ENV{GROFF_TMPDIR} if exists($ENV{GROFF_TMPDIR});

while (my $c=shift)
{
    $c=~s/(?<!\\)"/\\"/g;
    
    if (substr($c,0,2) eq '-T')
    {
	if (length($c) > 2)
	{
	    $dev=substr($c,2);
	}
	else
	{
	    $dev=shift;
	}
	next;
    }
    elsif (substr($c,0,2) eq '-K')
    {
	if (length($c) > 2)
	{
	    $preconv=$c;
	}
	else
	{
	    $preconv=$c;
	    $preconv.=shift;
	}
	next;
    }
    elsif (substr($c,0,2) eq '-k')
    {
	$preconv=$c;
	next;
    }
    elsif ($c eq '-z' or $c eq '-Z')
    {
	$dev=$c;
	next;
    }
    elsif ($c eq '-v' or $c eq '--version')
    {
	print "GNU pdfmom (groff) version @VERSION@\n";
	exit;
    }
    elsif (substr($c,0,1) eq '-')
    {
	if (length($c) > 1)
	{
	    push(@cmd,"\"$c\"");
	    push(@cmd,"'".(shift)."'") if length($c)==2 and index('dDfFIKLmMnoPrwW',substr($c,-1)) >= 0;
	}
	else
	{
	    # Just a '-'
	    
	    push(@cmd,$c);
	    $readstdin=2;
	}
    }
    else
    {
	# Got a filename?
	
	push(@cmd,"\"$c\"");
	$readstdin=0 if $readstdin == 1;
	
    }

}

my $cmdstring=' '.join(' ',@cmd).' ';

if ($readstdin)
{
    my ($fh,$tmpfn)=tempfile('pdfmom-XXXXX', UNLINK=>1);

    while (<STDIN>)
    {
	print $fh ($_);
    }
    
    close($fh);
    
    $cmdstring=~s/ - / $tmpfn / if $readstdin == 2;
    $cmdstring.=" $tmpfn " if $readstdin == 1;
}

if ($dev eq 'pdf')
{
    system("groff -Tpdf -dLABEL.REFS=1 -mom -z $cmdstring 2>&1 | LC_ALL=C grep '^\\. *ds' | groff -Tpdf -dPDF.EXPORT=1 -dLABEL.REFS=1 -mom -z - $cmdstring 2>&1 | LC_ALL=C grep '^\\. *ds' | groff -Tpdf -mom $preconv - $cmdstring");
}
elsif ($dev eq 'ps')
{
    system("groff -Tpdf -dLABEL.REFS=1 -mom -z $cmdstring 2>&1 | LC_ALL=C grep '^\\. *ds' | pdfroff -mpdfmark -mom --no-toc - $preconv $cmdstring");
}
elsif ($dev eq '-z') # pseudo dev - just compile for warnings
{
    system("groff -Tpdf -mom -z $cmdstring");
}
elsif ($dev eq '-Z') # pseudo dev - produce troff output
{
    system("groff -Tpdf -mom -Z $cmdstring");
}
else
{
    print STDERR "Not compatible with device '-T $dev'\n";
    exit 1;
}

# Local Variables:
# fill-column: 72
# mode: CPerl
# End:
# vim: set cindent noexpandtab shiftwidth=2 softtabstop=2 textwidth=72:
